File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    answer = 0
    N = len(jobs)
    heap = []
    count = 0
    start = -1
    current = 0
    
    while count < N:
        for j in jobs:
            if start < j[0] <= current:
                heapq.heappush(heap, [j[1], j[0]])
        
        if heap:
            [x, y] = heapq.heappop(heap)
            start = current
            current += x
            answer += current - y
            count += 1
        else:
            current += 1

    return answer // N

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int finish = 0;
        int answer = 0;
        
        Arrays.sort(jobs, (x, y) -> x[0] != y[0] ? x[0] - y[0] : x[1] - y[1]);
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((x, y) -> x[1] - y[1]);
        int i = 0;
        
        while (i < jobs.length) {
            //System.out.println(finish);
            if (pq.isEmpty() && jobs[i][0] >= finish) {
                finish = jobs[i][1] + jobs[i][0];
                answer += jobs[i][1];
                i++;
                continue;
            }
            
            if (jobs[i][0] <= finish) {
                pq.offer(new int[]{jobs[i][0], jobs[i][1]});
                i++;
                continue;
            }
            
            if (!pq.isEmpty()) {
                int[] temp = pq.poll();
                answer += (finish - temp[0]) + temp[1];
                finish += temp[1];
            }
        }
        
        while (!pq.isEmpty()) {
            int[] temp = pq.poll();
            answer += (finish - temp[0]) + temp[1];
            finish += temp[1];
        }
        
        System.out.println(finish);
        
        return answer / jobs.length;
    }
}